pkg/filter/and.go

```diff
@@ -0,0 +1,111 @@
+package filter
+
+import (
+	"fmt"
+	"sort"
+)
+
+// AllocationFilterOr is a set of filters that should be evaluated as a logical
+// AND.
+type And[T any] struct {
+	Filters []Filter[T]
+}
+
+func (a And[T]) String() string {
+	s := "(and"
+	for _, f := range a.Filters {
+		s += fmt.Sprintf(" %s", f)
+	}
+
+	s += ")"
+	return s
+}
+
+func (a And[T]) Matches(that T) bool {
+	filters := a.Filters
+	if len(filters) == 0 {
+		return true
+	}
+
+	for _, filter := range filters {
+		if !filter.Matches(that) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Flattened converts a filter into a minimal form, removing unnecessary
+// intermediate objects
+//
+// Flattened returns:
+// - AllPass if filter contains no filters
+// - AllCut if filter contains AllCut
+// - the inner filter if filter contains one filter
+// - an equivalent AllocationFilterAnd if filter contains more than one filter
+func (a And[T]) Flattened() Filter[T] {
+	var flattenedFilters []Filter[T]
+	for _, innerFilter := range a.Filters {
+		flattenedInner := innerFilter.Flattened()
+
+		// Ignore AllPass in And
+		if flattenedInner.equals(AllPass[T]{}) {
+			continue
+		}
+
+		// AllCut means the entire And is AllCut
+		if flattenedInner.equals(AllCut[T]{}) {
+			return AllCut[T]{}
+		}
+
+		flattenedFilters = append(flattenedFilters, flattenedInner)
+	}
+
+	// Remove duplicates
+	flattenedFilters = removeDuplicates(flattenedFilters)
+
+	// Check for negations, if they exist And becomes AllCut
+	if checkForNegations(flattenedFilters) {
+		return AllCut[T]{}
+	}
+
+	// Empty And is an AllPass
+	if len(flattenedFilters) == 0 {
+		return AllPass[T]{}
+	}
+
+	if len(flattenedFilters) == 1 {
+		return flattenedFilters[0]
+	}
+
+	// If more than half of the children are Not filters And should have DeMorgans rule applied to it
+	if percentNot(flattenedFilters) > 0.5 {
+		return Not[T]{Or[T]{negateFilters(flattenedFilters)}}
+	}
```

Comment:
@michaelmdresser I had to write out some basic examples to test this:

Assuming that the variables in the following are a "push" (just a bool), you can count how many resulting operations there are. This _also_ assumes all operations have equal weight:

### 2 Nots out of 3

```
!a && !b && c
```
Ops: `And`, `Not`, `Not`
Total: `3`

Applying the 50% Not rule:

```
!(a || b || !c)
```
Ops: `Not`, `Or`, `Not`
Total: `3`

These are equivalent in total ops: 🚫 

---- 

### 3 Nots out of 4

```
!a && !b && !c && d
```
Ops: `And`, `Not`, `Not`, `Not`
Total: `4`

Applying the 50% Not rule:

```
!(a || b || c || !d)
```
Ops: `Not`, `Or`, `Not`
Total: `3`

Applying DeMorgan's yields 1 less operation: ✔️ 

---- 

### 3 Nots out of 5

```
!a && !b && !c && d && e
```
Ops: `And`, `Not`, `Not`, `Not`
Total: `4`

Applying the 50% Not rule:

```
!(a || b || c || !d || !e)
```
Ops: `Not`, `Or`, `Not`, `Not`
Total: `4`

These are equivalent in total ops: 🚫 

----

It looks like the `test` to see if a combination of filters can be reduced should be: 
```go
totalOps := // ...
totalNotOps := // ... 
remainingOps := totalOps - totalNotOps  

// Because we are adding +1 NotOp in exchange for TotalNotOps, but that involves flipping RemainingOps, 
// TotalNotOps has to be > RemainingOps + 1
if totalNotOps > remainingOps + 1 { 
    return Not[T]{Or[T]{negateFilters(flattenedFilters)}}
}
```

Reply:
Now that I'm following a little closer, I see that the point is to prove equivalency. The only concern I have is:

`!(a || b || !c || !d)` == `(!a && !b && c && d)` when flattening would become:

`!!(!a && !b && c && d)` == `(!a && !b && c && d)`

Do we recursively simplify here? Does the `!!` cancel?

Reply:
I think the !! does cancel, but check `Not[T] Flatten()` to be sure

Reply:
I mean, I see that `Not[T].Flatten()` does remove double bang, but what we return from the DeMorgan's rule is:
```go
return Not[T]{Or[T]{negateFilters(flattenedFilters)}}
```

I think we would have to return: 
```go
return Not[T]{Or[T]{negateFilters(flattenedFilters)}}.Flatten()
```

in order to be equivalent.